### PR TITLE
[MLIR][MathDialect] fix fp32 promotion crash when encounters scf.if

### PR DESCRIPTION
**Pull request by crazydemo**
1. Expand legal op list in `legalizeToF32`
2. add legalization support for `math::rsqrtOp` in `mathToLibm`.

### CODE REVIEW

**Comment by crazydemo** [comment]
take the below case as an example:
```
func.func @promote_in_if_block(%arg0: bf16, %arg1: bf16, %arg2: i1) -> bf16 {
  %0 = scf.if %arg2 -> bf16 {
    %1 = math.absf %arg0 : bf16
    scf.yield %1 : bf16
  } else {
    scf.yield %arg1 : bf16
  }
  return %0 : bf16
}
```

after applied `LegalizeToF32RewritePattern` will get:
```
"builtin.module"() ({
  "func.func"() <{function_type = (bf16, bf16, i1) -> bf16, sym_name = "promote_in_if_block"}> ({
  ^bb0(%arg0: bf16, %arg1: bf16, %arg2: i1):
    %0 = "scf.if"(%arg2) : (i1) -> f32
    %1 = "arith.truncf"(%0) <{fastmath = #arith.fastmath<contract>}> : (f32) -> bf16
    %2 = "scf.if"(%arg2) ({
      "scf.yield"(%arg0) : (bf16) -> ()
    }, {
      "scf.yield"(%arg1) : (bf16) -> ()
    }) : (i1) -> bf16
    "func.return"(%2) : (bf16) -> ()
  }) : () -> ()
}) : () -> ()
```
Note that here's a newly generated `%0 = "scf.if"(%arg2) : (i1) -> f32`, which is invalid for further scf::IfOp folder, as its `numRegions=1`. The folder function needs the `numRegions=2`, otherwise when `getElseRegion()` will exceed the index range.

**Review by krzysz00 — CHANGES_REQUESTED** [review]
I think this change could be generalized and introduces a dependency on `scf` that's not needed here - that is, this could cause issues with, say, `scf.for`

File: mlir/lib/Dialect/Math/Transforms/LegalizeToF32.cpp [hunk 1]
```diff
@@ -69,6 +70,7 @@ void mlir::math::populateLegalizeToF32ConversionTarget(
       [&typeConverter](Operation *op) -> bool {
         return typeConverter.isLegal(op);
       });
+  target.addLegalOp<scf::IfOp>();
```
[the remark this answers is not in the document]
Comment (crazydemo) [reply mlir/lib/Dialect/Math/Transforms/LegalizeToF32.cpp]: I set all the `non-math` dialects as legal. Maybe Arith should be processed in `emulate-unsupported-floats`.

File: mlir/lib/Conversion/MathToLibm/MathToLibm.cpp [hunk 2]
```diff
@@ -191,6 +191,7 @@ void mlir::populateMathToLibmConversionPatterns(RewritePatternSet &patterns) {
   populatePatternsForOp<math::SinOp>(patterns, ctx, "sinf", "sin");
   populatePatternsForOp<math::SinhOp>(patterns, ctx, "sinhf", "sinh");
   populatePatternsForOp<math::SqrtOp>(patterns, ctx, "sqrtf", "sqrt");
+  populatePatternsForOp<math::RsqrtOp>(patterns, ctx, "rsqrtf", "rsqrt");
```
Comment (crazydemo) [inline mlir/lib/Conversion/MathToLibm/MathToLibm.cpp:194]: find `rsqrtOp` is not legalized in `mathToLibm pass`, fix it in this PR.

**Review by krzysz00 — APPROVED** [review]
Lgtm, though maybe break the `rsqrt` change into a different PR

**Review by BRUCE11111 — APPROVED** [review]

**Pull request merged**